Add Tensor Parallel Tutorial (#2829)
pytorch/tutorials: intermediate_source/TP_tutorial.rst

Large Scale Transformer Model training with Tensor Parallel (TP)
======================================================

> **Author**: `Wanchao Liang<https://github.com/wanchaol>`__, `Tianyu Liu<https://github.com/tianyu-l>`__,

Reviewer: ```suggestion
**Author**: `Wanchao Liang <https://github.com/wanchaol>`__, `Tianyu Liu <https://github.com/tianyu-l>`__,
```

.. figure:: /_static/img/distributed/megatron_lm.png
   :width: 100%
   :align: center
   :alt: Megatron-LM TP

   Figure 1. represents the sharding in Tensor Parallelism style on a Transformer model’s MLP and Self-Attention layer, where the matmuls in both attention/MLP happens through sharded computations (`image source <https://arxiv.org/abs/1909.08053>__`)

At a high level, PyTorch Tensor Parallel works as follows:

*Sharding initialization*

* Determine which ``ParallelStyle`` to apply to each layer and shard the initialized module by calling ``parallelize_module``
* The parallelized modules would have their model parameters be swapped to DTensors, and DTensor would be responsible to run the parallelized module.

*Runtime foward/backward*

> * Depending on the input/outputs DTensor layouts user specified for each parallel style, it would run proper communication operation to transform the DTensor layouts for inputs/outputs

Reviewer 2: This bullet item was not properly rendered. Maybe need a blank line before?

The world size (# GPUs) becoming too large (> 128/256 GPUs) and the FSDP collectives (i.e. ``allgather``) are being dominated by ring latency. TP/SP on top of FSDP would cut the FSDP world size by 8 thereby reducing the latency costs by 8.
Hit data parallelism limit where you can not raise the global batch size to be above the # GPUs due to both convergence and GPU memory limitations, Tensor/Sequence Parallelism is the only known way to “ballpark” the global batch size and continue scaling with more GPUs. This means both model size and number of GPUs could continue to scale.
For certain types of models, when local batch size becomes smaller TP/SP gives more flop friendly matrix multiplication shapes.

So when pre-training, how easy is it to hit those limits? Today to pre-train a Large Language Model (LLM) with billions/trillions of tokens it could take months even with thousands of GPUs.

> It will always hit limitation 1 when training LLM on a large scale. For example, Llama 2 70B trained with 2k GPUs for 35 days, multi-dimensional parallelisms are needed at 2k scale.

Reviewer 2: Here we are referring to **limitation 1** and **limitation 2**, shall we consider using numbered points for the limitations above?

When the Transformer model becomes larger (i.e. Llama2 70B), it will also quickly hit the limitation 2: One could not use FSDP alone with even local batch_size=1 due to memory and convergence constraints. For example, Llama 2 global batch size is 1k, so data parallelism alone can not be used at 2k GPUs.

How to apply Tensor Parallel
---------------

To demonstrate how to apply tensor parallelism using the PyTorch native Tensor Parallel APIs, let us look at a common Transformer model. In this tutorial, we use the most recent `Llama2 model <https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/>__` as a reference Transformer model implementation, as it is also widely used in the community.

> Since Tensor Parallel shards individual tensors over a set of devices, we would need to set up the distributed environment (i.e NCCL communicators) first. Tensor Parallelism is a SPMD sharding algorithm similar to PyTorch DDP/FSDP, and it under the hood leverages the PyTorch DTensor to perform sharding. It also utilizes the DeviceMesh abstraction (which under the hood manages ProcessGroups) for device management and sharding. To see how to utilize DeviceMesh to set up multi-dimensional parallelisms, please refer to this tutorial. Tensor Parallel usually works within each host, so let us first initialize a DeviceMesh that connects 8 GPUs within a host.

Reviewer: ```suggestion
Since Tensor Parallel shards individual tensors over a set of devices, we would need to set up the distributed environment (such as NCCL communicators) first. Tensor Parallelism is a Single-Program Multiple-Data (SPMD) sharding algorithm similar to PyTorch DDP/FSDP, and it under the hood leverages the PyTorch DTensor to perform sharding. It also utilizes the DeviceMesh abstraction (which under the hood manages ProcessGroups) for device management and sharding. To see how to utilize DeviceMesh to set up multi-dimensional parallelisms, please refer to `this tutorial <https://pytorch.org/tutorials/recipes/distributed_device_mesh.html>`__. Tensor Parallel usually works within each host, so let us first initialize a DeviceMesh that connects 8 GPUs within a host.
```

> .. code-block:: python

Reviewer: there needs to be a blank line between ``.. code-block:: python`` and the code/coment. Otherwise, the code block will not render in HTML.

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    # run this via torchrun: torchrun --standalone --nproc_per_node=8 ./tp_tutorial.py

    from torch.distributed.device_mesh import init_device_mesh

    tp_mesh = init_device_mesh("cuda", (8,))

> Now we initialize the DeviceMesh, let us take a detailed look at the Llama2 model architecture and see how we should perform the tensor parallel sharding. Here we focus on the core TransformerBlock, where the Transformer model stacks the identical TransformerBlocks to scale up the model.

Reviewer: ```suggestion
Now that we have initialized DeviceMesh, let us take a detailed look at the Llama2 model architecture and see how we should perform the tensor parallel sharding. Here we focus on the core TransformerBlock, where the Transformer model stacks the identical TransformerBlocks to scale up the model.
```

> The core TransformerBlock consists of an Attention layer and a FeedForward layer. Let us first look at the simpler FeedForward layer. For the FeedForward Layer it consists of three Linear layers, where it performs a SwiGLU style MLP, looking at its forward function:

Reviewer: I've found different ways of writing "Feed Forward". I suggest we stick to Feed Forward unless there is a reason to write it as one word.
```suggestion
The core ``TransformerBlock`` consists of an Attention layer and a Feed Forward layer. Let us first look at the simpler Feed Forward layer. For the Feed Forward Layer it consists of three Linear layers, where it performs a SwiGLU style MLP, looking at its forward function:
```

Author: Oh I think FeedForward is specifically a class in the Transformer model code of Lllama 2, so I think let's keep the term and quote it  with ``

.. code-block:: python

    def forward(self, x):
        return self.w2(F.silu(self.w1(x)) * self.w3(x))

> It performs ``w1`` and ``w3`` matmuls concurrently and followed by a w2 matmul with the result of the combined w1/w3 linear projection results. This means we could use the idea from the Tensor Parallelism paper to shard the w1/w3 Linear layers in the colwise fashion and shard the w2 Linear layer in the rowwise fashion, so that there iss only one allreduce communication happening at the end of all the three layers. With the PyTorch native Tensor Parallel, we can simply create a parallelize_plan for the FeedForward layer like below:

Reviewer: ```suggestion
It performs ``w1`` and ``w3`` matmuls concurrently and followed by a w2 matmul with the result of the combined w1/w3 linear projection results. This means we could use the idea from the Tensor Parallelism paper to shard the w1/w3 Linear layers in the colwise fashion and shard the w2 Linear layer in the rowwise fashion, so that there is only one ``allreduce`` communication happening at the end of all the three layers. With the PyTorch native Tensor Parallel, we can simply create a ``parallelize_plan`` for the Feed Forward layer like below:
```

Reviewer 2: pls consider making naming consistent: e.g. w2 -> \`\`w2\`\`

.. code-block:: python

    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module

    tp_plan = {
        # by default ColwiseParallel input layouts is replicated
        # and RowwiseParallel output layouts is replicated
        "feed_foward.w1": ColwiseParallel(),
        "feed_forward.w2": RowwiseParallel(),
        "feed_forward.w3": ColwiseParallel(),
    }

> That's simply how we configure the shardings for the FeedForward layer using the PyTorch Tensor Parallel APIs. Note that users would only need to specify how to shard the individual layers and the communications (i.e. allreduce) will happen under the hood.

Reviewer: ```suggestion
That's simply how we configure the shardings for the Feed Forward layer using the PyTorch Tensor Parallel APIs. Note that users would only need to specify how to shard the individual layers and the communications (for example, ``allreduce``) will happen under the hood.
```

> Next for the Attention Layer, it consists of ``wq, wk, wv`` Linear layers to project input to q/k/v, and then it performs attention and output projection with the “wo” Linear layer. Tensor Parallelism here intended to perform colwise sharding for the q/k/v projection and rowwise sharding for the wo linear projection. So we can add the Attention plan to the tp_plan that we just drafted up:

Reviewer: ```suggestion
Moving on to the Attention Layer. It consists of ``wq``, ``wk``, ``wv`` Linear layers to project input to q/k/v, and then it performs attention and output projection with the ``wo`` Linear layer. Tensor Parallelism here intended to perform column-wise  sharding for the q/k/v projection and row-wise sharding for the ``wo`` linear projection. So we can add the Attention plan to the ``tp_plan`` that we just drafted up:
```

Reviewer 2: intended -> intends

.. code-block:: python

    tp_plan = {
        # by default ColwiseParallel input layouts is replicated
        # and RowwiseParallel output layouts is replicated
        "attention.wq": ColwiseParallel(),
        "attention.wk": ColwiseParallel(),
        "attention.wv": ColwiseParallel(),
        "attention.wo": RowwiseParallel(),
        "feed_foward.w1": ColwiseParallel(),
        "feed_forward.w2": RowwiseParallel(),
        "feed_forward.w3": ColwiseParallel(),
    }

> This is almost the ``tp_plan`` we need to apply Tensor Parallelism to the llama model, however one thing we should be aware is that when sharding the linear layer colwise, the output of the linear layers would become sharded on the last tensor dimension, and the rowwise sharding linear layer directly accepts an input that shards on the last dimension. If there’re any more tensor operations (i.e. view operations) between the colwise linear and the rowwise linear, we would need to adjust the relevant shape related ops to sharded shape.

Reviewer: ```suggestion
This is almost the ``tp_plan`` we need to apply Tensor Parallelism to the Llama model. However, one thing we should be aware is that when sharding the linear layer column-wise, the output of the linear layers would become sharded on the last tensor dimension, and the row-wise sharding linear layer directly accepts an input that shards on the last dimension. If there are any more tensor operations (such as view operations) between the column-wise linear and the row-wise linear, we would need to adjust the relevant shape related ops to sharded shape.
```

> For the Llama model, in the attention layer there are couple of view operations that are shape related, the main source is that after colwise parallel for wq/wk/wv linear layers, the activation is sharded on the num_heads dimension, so we would need to adjust the num_heads to local num_heads.

Reviewer: ```suggestion
For the Llama model, in the attention layer there are couple of view operations that are shape related, the main source is that after column-wise parallel for ``wq``/``wk``/``wv`` linear layers, the activation is sharded on the ``num_heads`` dimension, so we would need to adjust the ``num_heads`` to local ``num_heads``.
```

Reviewer 2: there are couple of -> there are **a** couple of

Reviewer 2: Please break this into two sentences.

> Finally, we need to call ``parallelize_module`` API to make the plan for each TransformerBlock effective. Under the hood, it distributes the model parameters inside Attention and FeedForward layers to DTensors, and registers communication hooks for model inputs and outputs (before and after each module respectively), if necessary:

Reviewer: ```suggestion
Finally, we need to call ``parallelize_module`` API to make the plan for each ``TransformerBlock`` effective. Under the hood, it distributes the model parameters inside Attention and Feed Forward layers to DTensors, and registers communication hooks for model inputs and outputs (before and after each module respectively), if necessary:
```

.. code-block:: python

    for layer_id, transformer_block in enumerate(model.layers):
        layer_plan = {...}

        # Adjust attention module to use the local number of heads
        attn_layer = transformer_block.attention
        attn_layer.n_heads = attn_layer.n_heads // tp_mesh.size()
        attn_layer.n_kv_heads = attn_layer.n_kv_heads // tp_mesh.size()

        parallelize_module(
            module=transformer_block,
            device_mesh=tp_mesh,
            parallelize_plan=layer_plan,
        )

> Now that we elaborates the sharding plan for each TransformerBlock, there is usually a ``nn.Embedding`` in the first layer and a final ``nn.Linear`` projection layer, where user could choose rowwise/colwise sharding to the first nn.Embedding and colwise sharding to the last project layer with proper input/output layouts specified.

Reviewer: ```suggestion
Now that we have elaborated the sharding plan for each ``TransformerBlock``, there is usually a ``nn.Embedding`` in the first layer and a final ``nn.Linear`` projection layer, where user could choose row-wise or column-wise sharding to the first ``nn.Embedding`` and column-wise sharding to the last project layer with proper input and output layouts specified.
```

> NOTE: If the model to be partitioned is too large to fit into CPU memory, one could either use “meta” device initialization (i.e. initialize the model on meta device first, shard the layers, and the materialize the model), or parallelize the TransformerBlock layer by layer during the Transformer model initialization.

Reviewer: ```suggestion
.. note::
   If the model to be partitioned is too large to fit into CPU memory, one could either use ``meta`` device initialization (for example, initialize the model on meta device first, shard the layers, and the materialize the model), or parallelize the ``TransformerBlock`` layer by layer during the Transformer model initialization.
```

How to apply Sequence Parallel on ``LayerNorm/RMSNorm`` layers

> -----------

Reviewer: please make as long as the text above

> Sequence Parallel works on top of the Tensor Parallel illustrated above. Compared with basic Tensor Parallel, which only shards tensors within the Attention modules and FeedForward modules and keep their module inputs and outputs (namely activations in the forward pass and gradients in the backward pass) replicated, Sequence Parallel keeps them sharded on the sequence dimension.

Reviewer: ```suggestion
Sequence Parallel works on top of the Tensor Parallel illustrated above. Compared with basic Tensor Parallel, which only shards tensors within the Attention modules and Feed Forward modules and keeps their module inputs and outputs (namely activations in the forward pass and gradients in the backward pass) replicated, Sequence Parallel keeps them sharded on the sequence dimension.
```

> In a typical TransformerBlock, the forward function combines norm layers (``LayerNorm`` or ``RMSNorm``), an attention layer, a feedforward layer, and residual connections. For example:

Reviewer: ```suggestion
In a typical ``TransformerBlock``, the forward function combines norm layers (``LayerNorm`` or ``RMSNorm``), an attention layer, a Feed Forward layer, and residual connections. For example:
```

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    # forward in a TransformerBlock
    def forward(self, x):
        h = x + self.attention(self.attention_norm(x))
        out = h + self.feed_forward(self.ffn_norm(h))
        return out

> In most use cases, the activations (and gradients) are of the shape ``[batch size, sequence length, hidden dimension]`` outside the Attention and FeedForward modules. In the DTensor’s language, Sequence Parallelism maintains a Shard(1) layout for the activations, and does conversions as needed when entering or exiting those modules. Following the code example earlier, the code below demonstrates how we apply Sequence Parallel to the norm layers within a TransformerBlock:

Reviewer: ```suggestion
In most use cases, the activations (and gradients) are of the shape ``[batch size, sequence length, hidden dimension]`` outside the Attention and Feed Forward modules. In the DTensor’s language, Sequence Parallelism maintains a Shard(1) layout for the activations, and does conversions as needed when entering or exiting those modules. Following the code example earlier, the code below demonstrates how we apply Sequence Parallel to the norm layers within a TransformerBlock:
```

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    from torch.distributed.tensor.parallel import (
        ColwiseParallel,
        PrepareModuleInput,
        RowwiseParallel,
        SequenceParallel,
        parallelize_module
    )

    tp_plan = {
        # by default the input and output of SequenceParallel
        # has Shard(1) layouts
        "attention": PrepareModuleInput(
            input_layouts=(Shard(1),),
            desired_input_layouts=(Replicate(),),
        ),
        "attention.wq": ColwiseParallel(),
        "attention.wk": ColwiseParallel(),
        "attention.wv": ColwiseParallel(),
        "attention.wo": RowwiseParallel(output_layouts=Shard(1)),
        ​​"attention_norm": SequenceParallel(),
        "feed_forward": PrepareModuleInput(
            input_layouts=(Shard(1),),
            desired_input_layouts=(Replicate(),),
        ),
        "feed_foward.w1": ColwiseParallel(),
        "feed_forward.w2": RowwiseParallel(output_layouts=Shard(1)),
        "feed_forward.w3": ColwiseParallel(),
        "ffn_norm": SequenceParallel(),
    }

> One can see we now use PrepareModuleInput to modify the module input layouts to the Attention and FeedForward layers from ``Shard(1)`` to ``Replicate()``, and mark their output layouts as ``Shard(1)``. Just like what happens to Tensor Parallelism, one only needs to specify the tensor sharding layouts of the inputs and outputs, and the communication between layers will happen automatically.

Reviewer: ```suggestion
One can see we now use ``PrepareModuleInput`` to modify the module input layouts to the Attention and Feed Forward layers from ``Shard(1)`` to ``Replicate()``, and mark their output layouts as ``Shard(1)``. Just like what happens to Tensor Parallelism, one only needs to specify the tensor sharding layouts of the inputs and outputs, and the communication between layers will happen automatically.
```

> Note that with Sequence Parallel, we assume the inputs and outputs of a TransformerBlock are always sharded on the sequence dimension, so that multiple TransformerBlocks can be concatenated seamlessly. The only exception is that the input to the first TransformerBlock is replicated, so it has to be converted explicitly:

Reviewer: ```suggestion
Note that with Sequence Parallel, we assume the inputs and outputs of a ``TransformerBlock`` are always sharded on the sequence dimension, so that multiple ``TransformerBlocks`` can be concatenated seamlessly. The only exception is that the input to the first ``TransformerBlock`` is replicated, so it has to be converted explicitly:
```

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

.. figure:: /_static/img/distributed/loss_parallel.png
   :width: 100%
   :align: center
   :alt: loss parallel

   Figure 2. Cross-entropy loss forward computation with loss parallel on one GPU. Blue represents sharded tensors; green represents replicated tensors; yellow represents tensors with partial values (to be all-reduced). Black arrows are local computations; red arrows are functional collectives among GPUs.

In the PyTorch Tensor Parallel API, Loss Parallel can be enabled via a context manager ``loss_parallel``, with which one can directly use ``torch.nn.functional.cross_entropy`` or ``torch.nn.CrossEntropyLoss`` without modifying other parts of their code.

To apply Loss Parallel, the model predictions, usually of the shape ``[batch size, sequence length, vocabulary size]``, should be sharded on the vocabulary dimension. This can be easily done via marking the output layouts of the last linear projection layer output:

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    model = parallelize_module(
        model,
        tp_mesh,
        {
            "output": ColwiseParallel(
                input_layouts=Shard(1),
                use_local_output=False,
            ),
            "norm": SequenceParallel(),
            "layers.0": PrepareModuleInput(
                input_layouts=(Replicate(),),
                desired_input_layouts=(Shard(1),),
            ),
        },
    )

In the code above, we also apply Sequence Parallel to the norm layer before output. We apply ``use_local_output=False`` to let the output stay as a DTensor, to work with the ``loss_parallel`` context manager. After that, one can simply call the cross_entropy loss function as is shown below. Note that the backward computation also needs to happen within the context.

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    import torch.nn.functional as F
    from torch.distributed.tensor.parallel import loss_parallel

    pred = model(input_ids)
    with loss_parallel():
        # assuming pred and labels are of the shape [batch, seq, vocab]
        loss = F.cross_entropy(pred.flatten(0, 1), labels.flatten(0, 1))
        loss.backward()

How to combine Tensor Parallel with Fully Sharded Data Parallel together
------------

> Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together. Since Tensor Parallelism incurs communications that’s blocking the computation, we want to make sure it runs within a fast communication channel (i.e. NVLink). In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.

Reviewer: ```suggestion
Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together. Since Tensor Parallelism incurs communications that’s blocking the computation, we want to make sure it runs within a fast communication channel, such as NVLink. In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.
```

.. figure:: /_static/img/distributed/fsdp_tp.png
   :width: 100%
   :align: center
   :alt: fsdp + tp
   Figure 3. FSDP and TP work on separate device dimensions, FSDP communication happens inter-host and TP communication happens intra-host.

This 2-D parallelism pattern can be easily expressed via a 2-D DeviceMesh, and we just need pass each “sub” DeviceMesh to each individual parallelism APIs:

> .. code-block:: python

Reviewer: ```suggestion
.. code-block:: python

```

    from torch.distributed.device_mesh import init_device_mesh
    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
    from torch.distributed.fsdp import FullyShardedDataParallel as FSDP

    # i.e. 2-D mesh is [dp, tp], training on 64 GPUs that performs 8 way DP and 8 way TP
    mesh_2d = init_device_mesh("cuda", (8, 8))
    tp_mesh = mesh_2d["tp"] # a submesh that connects intra-host devices
    dp_mesh = mesh_2d["dp"] # a submesh that connects inter-host devices

    model = Model(...)

    # apply tensor parallel
    model_tp = parallelize_module(model, tp_mesh, tp_plan)
    model_2d = FSDP(model_tp, device_mesh=dp_mesh, use_orig_params=True, ...)

This would allow us to easily apply Tensor Parallel within each host (intra-host) and apply FSDP across hosts (inter-hosts). The Tensor(Model) Parallelism and Data Parallelism techniques combined together provides the ability to continue increasing model size and efficient training using a large number of GPUs.

> We can see that the PyTorch Tensor Parallel makes 0-code changes to the Llama model, allowing fast enablement for prototyping and productionizations. To see the complete e2e code example about how to apply Tensor/Sequence Parallelism to the Llama model, please refer to the examples repository.

Reviewer: Please add a Conclusion. For example:

This tutorial demonstrates how to train a large Transformer-like model across hundreds to thousands of GPUs using Tensor Parallel in combination with Fully Sharded Data Parallel. It explains the working of Tensor Parallel, which is an efficient model parallelism technique to train large scale Transformer models. 

To learn more about Tensor Parallel, see:

* Link 1
* Link 2
* Link 3

.. code-block:: python

    import torch.nn.functional as F
    from torch.distributed.tensor.parallel import loss_parallel

    pred = model(input_ids)
    with loss_parallel():
        # assuming pred and labels are of the shape [batch, seq, vocab]
        loss = F.cross_entropy(pred.flatten(0, 1), labels.flatten(0, 1))
        loss.backward()

Combine Tensor Parallel with Fully Sharded Data Parallel together
-----------------------------------------------------------------

Now that we have shown how to apply Tensor/Sequence Parallel to the model, let us also take a look at how Tensor Parallel and Fully Sharded Data Parallel could work together.
Since Tensor Parallelism incurs communications that is blocking the computation, we want to make sure it runs within a fast communication channel, such as NVLink.
In practice, we usually apply Tensor Parallel within each host, and apply Fully Sharded Data Parallel across the hosts.

.. figure:: /_static/img/distributed/fsdp_tp.png
   :width: 100%
   :align: center
   :alt: fsdp + tp

>    Figure 3. FSDP and TP work on separate device dimensions, FSDP communication happens inter-host and TP communication happens intra-host.

Reviewer: ```suggestion

   Figure 3. FSDP and TP work on separate device dimensions, FSDP communication happens inter-host and TP communication happens intra-host.
```

Author: Thanks done!
